packages/noya-state/src/reducers/history.ts

```diff
@@ -0,0 +1,99 @@
+import produce from 'immer';
+import { ApplicationState, Action, reducer } from './application';
+import { createInitialInteractionState } from './interaction';
+
+export type HistoryEntry = {
+  actionType: Action[0];
+  timestamp: number;
+  state: ApplicationState;
+};
+
+export type HistoryState = {
+  past: HistoryEntry[];
+  present: ApplicationState;
+  future: HistoryEntry[];
+};
+
+export type HistoryAction = [type: 'undo'] | [type: 'redo'] | Action;
+
+const FILE_CHANGED_TIMEOUT = 300;
+
+export function historyReducer(state: HistoryState, action: HistoryAction) {
+  const currentPresent = state.present;
+  const [actionType] = action;
+  switch (actionType) {
+    case 'undo':
+      return produce(state, (state) => {
+        if (state.past.length > 0) {
+          const newPresent = state.past.pop() as HistoryEntry;
+          if (newPresent) {
+            state.future.unshift(
+              createHistoryEntry(newPresent.actionType, currentPresent),
+            );
+            state.present = newPresent.state;
+          }
+        }
+      });
+    case 'redo':
+      return produce(state, (state) => {
+        if (state.future.length > 0) {
+          const newPresent = state.future.shift();
+          if (newPresent) {
+            state.past.push(
+              createHistoryEntry(newPresent.actionType, currentPresent),
+            );
+            state.present = newPresent.state;
+          }
+        }
+      });
+    default:
+      const newPresent = reducer(currentPresent, action as Action);
+      return produce(state, (state) => {
+        const sketchFileChanged =
```

Comment:
This is going to be a really expensive comparison. Can we look into why `!==` isn't working? Maybe we can come up with something better

Reply:
Yeah, I was weary of using stringify since it's slow, I'll dig deeper on why we can't compare.